File: architecture-skills/design-patterns/examples/flyweight/flyweight.cpp
// Flyweight pattern in C++.
//
// std::shared_ptr<const TreeType> expresses both aspects of the pattern
// in the type: the intrinsic state is shared (shared_ptr) and immutable
// (const). Each Tree context carries only its coordinates plus a cheap
// pointer copy.

#include <iostream>
#include <memory>
#include <string>
#include <unordered_map>
#include <vector>

class TreeType {  // flyweight: immutable intrinsic state
public:
    TreeType(std::string name, std::string color)
        : name_(std::move(name)), color_(std::move(color)) {}

    void draw(int x, int y) const {
        std::cout << color_ << ' ' << name_ << " at (" << x << ", " << y << ")\n";
    }

private:
    std::string name_;
    std::string color_;
};

class TreeTypeFactory {
public:
    std::shared_ptr<const TreeType> get(const std::string& name, const std::string& color) {
        const std::string key = name + '/' + color;
        auto it = cache_.find(key);
        if (it == cache_.end()) {
            it = cache_.emplace(key, std::make_shared<const TreeType>(name, color)).first;
        }
        return it->second;
    }

    std::size_t size() const { return cache_.size(); }

private:
    std::unordered_map<std::string, std::shared_ptr<const TreeType>> cache_;
};

struct Tree {  // context: extrinsic state + shared flyweight
    int x;
    int y;
    std::shared_ptr<const TreeType> kind;
};

int main() {
    TreeTypeFactory factory;
    std::vector<Tree> forest = {
        {1, 1, factory.get("oak", "green")},
        {2, 4, factory.get("pine", "dark green")},
        {5, 2, factory.get("oak", "green")},
        {6, 6, factory.get("birch", "light green")},
        {8, 3, factory.get("pine", "dark green")},
        {9, 9, factory.get("oak", "green")},
    };

    for (const Tree& tree : forest) {
        tree.kind->draw(tree.x, tree.y);
    }

    std::cout << "trees planted: " << forest.size()
              << ", tree types in memory: " << factory.size() << '\n';
    return 0;
}

File: architecture-skills/design-patterns/examples/flyweight/flyweight.py
"""Flyweight pattern in Python.

Intrinsic state (tree species) is shared; extrinsic state (position)
lives in each context. Note: Python interns small ints and many strings
as built-in flyweights, and functools.lru_cache on a factory function
is the shortest explicit implementation — shown here.
"""

from __future__ import annotations

from dataclasses import dataclass
from functools import lru_cache


@dataclass(frozen=True)
class TreeType:
    """Flyweight: immutable intrinsic state shared by many trees."""

    name: str
    color: str

    def draw(self, x: int, y: int) -> None:
        print(f"{self.color} {self.name} at ({x}, {y})")


@lru_cache(maxsize=None)
def get_tree_type(name: str, color: str) -> TreeType:
    """Flyweight factory: same arguments always return the same instance."""
    return TreeType(name, color)


@dataclass
class Tree:
    """Context: extrinsic state plus a reference to the shared flyweight."""

    x: int
    y: int
    kind: TreeType


class Forest:
    def __init__(self) -> None:
        self._trees: list[Tree] = []

    def plant(self, x: int, y: int, name: str, color: str) -> None:
        self._trees.append(Tree(x, y, get_tree_type(name, color)))

    def draw(self) -> None:
        for tree in self._trees:
            tree.kind.draw(tree.x, tree.y)


if __name__ == "__main__":
    forest = Forest()
    forest.plant(1, 1, "oak", "green")
    forest.plant(2, 4, "pine", "dark green")
    forest.plant(5, 2, "oak", "green")
    forest.plant(6, 6, "birch", "light green")
    forest.plant(8, 3, "pine", "dark green")
    forest.plant(9, 9, "oak", "green")
    forest.draw()

    info = get_tree_type.cache_info()
    print(f"trees planted: 6, tree types in memory: {info.currsize}")
